libbpf-tools/tcppktlat.c

```diff
@@ -0,0 +1,262 @@
+		if (errno) {
+			fprintf(stderr, "invalid lport: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'r':
+		errno = 0;
+		env.rport = strtoul(arg, NULL, 10);
+		if (errno) {
+			fprintf(stderr, "invalid rport: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'w':
+		column_width = 26;
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		min_us = strtoll(arg, NULL, 10);
+		if (errno || min_us <= 0) {
+			fprintf(stderr, "Invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+		env.min_us = min_us;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static int handle_event(void *ctx, void *data, size_t data_sz)
+{
+	const struct event *e = data;
+	char saddr[48], daddr[48];
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	if (env.timestamp) {
+		time(&t);
+		tm = localtime(&t);
+		strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+		printf("%-8s ", ts);
+	}
+	inet_ntop(e->family, &e->saddr, saddr, sizeof(saddr));
+	inet_ntop(e->family, &e->daddr, daddr, sizeof(daddr));
+
+	printf("%-7d %-7d %-16s %-*s %-5d %-*s %-5d %-.2f\n",
+		e->pid, e->tid, e->comm, column_width, saddr, e->sport, column_width, daddr,
+		e->dport, e->delta_us / 1000.0);
+
+	return 0;
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	fprintf(stderr, "lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct bpf_buffer *buf = NULL;
+	struct tcppktlat_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	obj = tcppktlat_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->targ_pid = env.pid;
+	obj->rodata->targ_tid = env.tid;
+	obj->rodata->targ_sport = env.lport;
+	obj->rodata->targ_dport = env.rport;
+	obj->rodata->targ_min_us = env.min_us;
+
+	buf = bpf_buffer__new(obj->maps.events, obj->maps.heap);
+	if (!buf) {
+		err = -errno;
+		fprintf(stderr, "failed to create ring/perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	if (probe_tp_btf("tcp_probe")) {
+		bpf_program__set_autoload(obj->progs.tcp_probe, false);
+		bpf_program__set_autoload(obj->progs.tcp_rcv_space_adjust, false);
+	} else {
+		bpf_program__set_autoload(obj->progs.tcp_probe_btf, false);
+		bpf_program__set_autoload(obj->progs.tcp_rcv_space_adjust_btf, false);
+	}
+
+	err = tcppktlat_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to laod BPF object: %d\n", err);
```

Comment:
Since bpf_get_socket_cookie() only supported with recent kernel (>= 5.12?), maybe you could add some info in the error message to indicate one of potential reason is the kernel won't support helper bpf_get_socket_cookie?
I did not specify version number as this won't work for backported kernel.